frame/staking/src/lib.rs

```diff
@@ -3345,19 +3345,45 @@ impl<T: Config> Module<T> {
 	}
 }
 
-impl<T: Config> sp_election_providers::ElectionDataProvider<T::AccountId, T::BlockNumber>
+impl<T: Config> frame_election_provider_support::ElectionDataProvider<T::AccountId, T::BlockNumber>
 	for Module<T>
 {
-	fn desired_targets() -> u32 {
-		Self::validator_count()
+	fn desired_targets() -> data_provider::Result<(u32, Weight)> {
+		Ok((Self::validator_count(), <T as frame_system::Config>::DbWeight::get().reads(1)))
 	}
 
-	fn voters() -> Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)> {
-		Self::get_npos_voters()
+	fn voters(
+		maybe_max_len: Option<usize>,
+	) -> data_provider::Result<(Vec<(T::AccountId, VoteWeight, Vec<T::AccountId>)>, Weight)> {
+		// NOTE: reading these counts already needs to iterate a lot of storage keys, but they get
+		// cached. This is okay for the case of `Ok(_)`, but bad for `Err(_)`, as the trait does not
+		// report weight in failures.
```

Comment:
why cant it report some partial weight?

Reply:
For now this is not a user-triggered event, so it won't be a serious damage. Rest assured, I've documented this flaw and it will be treated: 

![Screenshot 2021-03-16 at 07 13 29](https://user-images.githubusercontent.com/5588131/111264047-c5c38c80-861e-11eb-9149-6f693aa3ebd9.png)


(Once this line of work is done I will convert all these notes to issues)

Reply:
For the purpose of this PR, I will assume worse case to keep you happy.